File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    void solve(int index,int target,vector<int> &candidates,vector<vector<int>>&res,vector<int>&temp){
		if(index==candidates.size()){
			if(target==0){
				res.push_back(temp);
			}
			return;
		}
		if(candidates[index]<=target){
			temp.push_back(candidates[index]);
			solve(index,target-candidates[index],candidates,res,temp);
			temp.pop_back();
		}
		solve(index+1,target,candidates,res,temp);
	}
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> temp;
        solve(0,target,candidates,res,temp);
        return res; 
    }
};

File: 39-combination-sum/39-combination-sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    const result = [];
    solve(0,target,[])
    function solve(index,target,temp){
        if(target==0){
            result.push(temp)
            return;
        }
        else if(target<0)
            return;
        for(let i=index;i<candidates.length;i++){
             solve(i,target-candidates[i],[...temp,candidates[i]])
        }
    }
    return result  
};